File: mirror/tier2-ruby/showcase/perf_caching.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Optimization
# Module: perf_caching.rb
# Purpose: Demonstrate caching and memoization

class CachedFibonacci
  def initialize
    @cache = {}
    @hits = 0
    @misses = 0
  end
  
  def compute(n)
    if @cache.key?(n)
      @hits += 1
      return @cache[n]
    end
    
    @misses += 1
    result = if n < 2
      n
    else
      compute(n - 1) + compute(n - 2)
    end
    
    @cache[n] = result
    result
  end
  
  def stats
    {
      hits: @hits,
      misses: @misses,
      cache_size: @cache.size
    }
  end
end

fib = CachedFibonacci.new
fib_10 = fib.compute(10)
fib_20 = fib.compute(20)
stats = fib.stats

result = {
  fib_10: fib_10,
  fib_20: fib_20,
  cache_hits: stats[:hits],
  cache_misses: stats[:misses],
  cache_enabled: true,
}

puts result.inspect

File: mirror/tier2-php/showcase/perf_caching.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Optimization
// Module: perf_caching.php

class CachedFibonacci {
    private $cache = [];
    private $hits = 0;
    private $misses = 0;
    
    public function compute(int $n): int {
        if (isset($this->cache[$n])) {
            $this->hits++;
            return $this->cache[$n];
        }
        
        $this->misses++;
        $result = $n < 2 ? $n : $this->compute($n - 1) + $this->compute($n - 2);
        
        $this->cache[$n] = $result;
        return $result;
    }
    
    public function getStats(): array {
        return [
            'hits' => $this->hits,
            'misses' => $this->misses,
            'cache_size' => count($this->cache),
        ];
    }
}

$fib = new CachedFibonacci();
$fib10 = $fib->compute(10);
$fib20 = $fib->compute(20);
$stats = $fib->getStats();

$result = [
    'fib_10' => $fib10,
    'fib_20' => $fib20,
    'cache_hits' => $stats['hits'],
    'cache_misses' => $stats['misses'],
    'cache_enabled' => true,
];

echo json_encode($result);
?>
